File: src/ui/format.ts
// Small formatting helpers for the renderer views.

import type { ItemReason } from "../shared/types";

/**
 * Compact relative time: "now", "5m", "3h", "2d", then a short date
 * ("Aug 12", with the year appended once it differs from the current one).
 */
export function relativeTime(iso: string, now: number = Date.now()): string {
  const then = Date.parse(iso);
  if (Number.isNaN(then)) {
    return "";
  }
  const seconds = Math.floor(Math.max(0, now - then) / 1000);
  if (seconds < 60) {
    return "now";
  }
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) {
    return `${minutes}m`;
  }
  const hours = Math.floor(minutes / 60);
  if (hours < 24) {
    return `${hours}h`;
  }
  const days = Math.floor(hours / 24);
  if (days < 7) {
    return `${days}d`;
  }
  const date = new Date(then);
  const options: Intl.DateTimeFormatOptions
    = date.getFullYear() === new Date(now).getFullYear()
      ? { month: "short", day: "numeric" }
      : { month: "short", day: "numeric", year: "numeric" };
  return date.toLocaleDateString(undefined, options);
}

/** How each normalized reason reads in a capsule. */
const REASON_LABELS: Record<string, string> = {
  mentioned: "mentioned",
  review_requested: "review requested",
  assigned: "assigned",
  approval_required: "approval required",
  author: "author",
  commented: "commented",
  subscribed: "subscribed",
  activity: "activity",
};

/**
 * Capsule text for an item's reason. The fallback covers items still carrying
 * a pre-normalization reason from the store; the next sync replaces them.
 */
export function reasonLabel(reason: ItemReason): string {
  return REASON_LABELS[reason] ?? reason.replace(/_/g, " ");
}

/** Idle sync label for the header; the syncing case is handled by the caller. */
export function syncedLabel(lastSyncAt: string | null, now: number = Date.now()): string {
  if (!lastSyncAt) {
    return "Not synced yet";
  }
  const rel = relativeTime(lastSyncAt, now);
  if (rel === "now") {
    return "Synced just now";
  }
  if (/^\d/.test(rel)) {
    return `Synced ${rel} ago`;
  }
  return `Synced ${rel}`;
}

File: Sources/Majordomo/UI/Format.swift
// Small formatting helpers for the views, ported from src/ui/format.ts.

import Foundation

private let sameYearFormatter: DateFormatter = {
  let f = DateFormatter()
  f.setLocalizedDateFormatFromTemplate("MMM d")
  return f
}()

private let otherYearFormatter: DateFormatter = {
  let f = DateFormatter()
  f.setLocalizedDateFormatFromTemplate("MMM d y")
  return f
}()

private let absoluteFormatter: DateFormatter = {
  let f = DateFormatter()
  f.dateStyle = .medium
  f.timeStyle = .short
  return f
}()

/// Compact relative time: "now", "5m", "3h", "2d", then a short date
/// ("Aug 12", with the year appended once it differs from the current one).
func relativeTime(_ date: Date, now: Date) -> String {
  let seconds = Int(max(0, now.timeIntervalSince(date)))
  if seconds < 60 {
    return "now"
  }
  let minutes = seconds / 60
  if minutes < 60 {
    return "\(minutes)m"
  }
  let hours = minutes / 60
  if hours < 24 {
    return "\(hours)h"
  }
  let days = hours / 24
  if days < 7 {
    return "\(days)d"
  }
  let calendar = Calendar.current
  let sameYear = calendar.component(.year, from: date) == calendar.component(.year, from: now)
  return (sameYear ? sameYearFormatter : otherYearFormatter).string(from: date)
}

/// "Sep 2, 2026, 10:41" in the user's locale, for the preview pane.
func absoluteTime(_ date: Date) -> String {
  absoluteFormatter.string(from: date)
}

/// An item's meta line — "repo · author · 1h", skipping missing parts —
/// shared by the list rows and the tray-menu rows.
func metaLine(for item: InboxItem, now: Date) -> String {
  var parts: [String] = []
  if !item.repo.isEmpty {
    parts.append(item.repo)
  }
  if let author = item.author {
    parts.append(author)
  }
  parts.append(relativeTime(item.updatedAt, now: now))
  return parts.joined(separator: " · ")
}
